ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/AbstractBlockParameterOrBlockHashMethod.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.JsonRpcRequestContext;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.parameters.BlockParameterOrBlockHash;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcError;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcErrorResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import org.hyperledger.besu.ethereum.api.query.BlockchainQueries;
+import org.hyperledger.besu.ethereum.core.Hash;
+
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
+
+public abstract class AbstractBlockParameterOrBlockHashMethod implements JsonRpcMethod {
+  private static final String INVALID_BLOCK_NUMBER_PARAMETER = "Invalid Block Number Parameter.";
+
+  protected final Supplier<BlockchainQueries> blockchainQueries;
+
+  protected AbstractBlockParameterOrBlockHashMethod(final BlockchainQueries blockchainQueries) {
+    this(Suppliers.ofInstance(blockchainQueries));
+  }
+
+  protected AbstractBlockParameterOrBlockHashMethod(
+      final Supplier<BlockchainQueries> blockchainQueries) {
+    this.blockchainQueries = blockchainQueries;
+  }
+
+  protected abstract BlockParameterOrBlockHash blockParameterOrBlockHash(
+      JsonRpcRequestContext request);
+
+  protected abstract Object resultByBlockNumber(JsonRpcRequestContext request, long blockNumber);
```

Comment:
I think this should be requestByBlockHash.  There can be multiple hashes at a particular height, only one of which is "cannonical".  We still want to be able to query non-cannonical blocks and those cannot be accessed by a block height lookup.

So where we have a number, do a lookup for the canonical hash.  And where we have a hash and do a number lookup, just pass in the number.

Or we need two methods and all children implement it twice.